File: apps/web/lib/ai.ts
export interface NegotiationMetadataPayload {
  contract_name?: string | null;
  contract_type?: string | null;
  counterparty?: string | null;
  effective_date?: string | null;
  renewal_date?: string | null;
  jurisdiction?: string | null;
  additional_context?: Record<string, unknown> | null;
}

export interface NegotiationContextPayload {
  topic: string;
  contract_id: string;
  template_id: string;
  vertical?: string | null;
  current_position: string;
  target_position: string;
  fallback_position?: string | null;
  stakeholders?: string[];
  impact?: string | null;
  risk_signal?: string | null;
  metadata?: NegotiationMetadataPayload | null;
}

export interface NegotiationRequestPayload {
  context: NegotiationContextPayload;
  prompt_override?: string | null;
}

export interface NegotiationGuidanceResponsePayload {
  summary: string;
  fallback_recommendation: string;
  talking_points?: string[];
  risk_callouts?: string[];
  confidence: number;
  cached?: boolean;
  model: string;
  latency_ms: number;
  documentation_url?: string | null;
  generated_prompt?: string | null;
}

export interface NegotiationGuidance {
  summary: string;
  fallbackRecommendation: string;
  talkingPoints: string[];
  riskCallouts: string[];
  confidence: number;
  cached: boolean;
  model: string;
  latencyMs: number;
  documentationUrl?: string | null;
  generatedPrompt?: string | null;
}

export class NegotiationApiError extends Error {
  status?: number;
  body?: unknown;

  constructor(message: string, options?: { status?: number; body?: unknown }) {
    super(message);
    this.name = 'NegotiationApiError';
    this.status = options?.status;
    this.body = options?.body;
  }
}

export interface RequestNegotiationGuidanceOptions {
  apiUrl?: string;
  promptOverride?: string | null;
  signal?: AbortSignal;
  fetchFn?: typeof fetch;
}

function resolveApiBaseUrl(): string | undefined {
  if (typeof window === 'undefined') {
    return process.env.CONTRACT_IQ_API_URL;
  }

  return process.env.NEXT_PUBLIC_CONTRACT_IQ_API_URL;
}

function normaliseGuidance(payload: NegotiationGuidanceResponsePayload): NegotiationGuidance {
  return {
    summary: payload.summary,
    fallbackRecommendation: payload.fallback_recommendation,
    talkingPoints: Array.isArray(payload.talking_points) ? payload.talking_points : [],
    riskCallouts: Array.isArray(payload.risk_callouts) ? payload.risk_callouts : [],
    confidence: payload.confidence,
    cached: Boolean(payload.cached),
    model: payload.model,
    latencyMs: payload.latency_ms,
    documentationUrl: payload.documentation_url ?? null,
    generatedPrompt: payload.generated_prompt ?? null
  };
}

async function parseErrorBody(response: Response): Promise<unknown> {
  const text = await response.text();
  if (!text) {
    return null;
  }

  try {
    return JSON.parse(text);
  } catch (error) {
    return text;
  }
}

export async function requestNegotiationGuidance(
  context: NegotiationContextPayload,
  options: RequestNegotiationGuidanceOptions = {}
): Promise<NegotiationGuidance> {
  const fetchImpl = options.fetchFn ?? fetch;
  const apiBase = options.apiUrl ?? resolveApiBaseUrl() ?? 'http://localhost:8000';

  const body: NegotiationRequestPayload = {
    context,
    prompt_override: options.promptOverride ?? undefined
  };

  const response = await fetchImpl(`${apiBase}/ai/negotiation`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(body),
    signal: options.signal,
    cache: 'no-store'
  } as RequestInit);

  if (!response.ok) {
    const errorBody = await parseErrorBody(response);
    let message: string;

    if (typeof errorBody === 'string' && errorBody.length > 0) {
      message = errorBody;
    } else {
      const detail = (errorBody as { detail?: unknown })?.detail;
      if (typeof detail === 'string' && detail.length > 0) {
        message = detail;
      } else {
        message = `Request failed with status ${response.status}`;
      }
    }

    throw new NegotiationApiError(message, { status: response.status, body: errorBody });
  }

  const payload = (await response.json()) as NegotiationGuidanceResponsePayload;
  return normaliseGuidance(payload);
}

File: apps/api/contract_iq/schemas/ai.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


class NegotiationMetadata(BaseModel):
    """Supplemental contract attributes that provide AI context."""

    model_config = ConfigDict(extra="allow")

    contract_name: str | None = Field(default=None, max_length=200)
    contract_type: str | None = Field(default=None, max_length=120)
    counterparty: str | None = Field(default=None, max_length=200)
    effective_date: datetime | None = Field(default=None)
    renewal_date: datetime | None = Field(default=None)
    jurisdiction: str | None = Field(default=None, max_length=120)
    additional_context: dict[str, Any] | None = Field(
        default=None,
        description="Arbitrary metadata that may influence guidance generation.",
    )


class NegotiationContext(BaseModel):
    """Key contract positions and signals used to generate negotiation guidance."""

    model_config = ConfigDict(extra="forbid")

    topic: str = Field(..., min_length=1, max_length=160, description="Clause or topic name.")
    contract_id: str = Field(..., min_length=1, description="Unique identifier for the contract.")
    template_id: str = Field(..., min_length=1, description="Template reference powering guidance.")
    vertical: str | None = Field(default=None, max_length=80, description="Industry vertical tag.")
    current_position: str = Field(
        ...,
        min_length=1,
        description="Current clause language or supplier position.",
    )
    target_position: str = Field(
        ...,
        min_length=1,
        description="Preferred or ideal counter-position for the customer.",
    )
    fallback_position: str | None = Field(
        default=None,
        min_length=1,
        description="Minimum acceptable fallback if the target cannot be achieved.",
    )
    stakeholders: list[str] = Field(
        default_factory=list,
        description="Stakeholders who should be informed about the negotiation guidance.",
    )
    impact: str | None = Field(
        default=None,
        max_length=64,
        description="Business impact level for the clause (e.g., High, Medium).",
    )
    risk_signal: str | None = Field(
        default=None,
        max_length=64,
        description="Risk signal classification (e.g., Elevated, Critical).",
    )
    metadata: NegotiationMetadata | None = Field(default=None)

    @field_validator("stakeholders", mode="before")
    @classmethod
    def _normalize_stakeholders(cls, value: Any) -> list[str]:
        if value is None:
            return []
        if isinstance(value, str):
            return [value.strip()] if value.strip() else []
        return [str(item).strip() for item in value if str(item).strip()]


class NegotiationRequest(BaseModel):
    """Request payload for AI-generated negotiation guidance."""

    model_config = ConfigDict(extra="forbid")

    context: NegotiationContext
    prompt_override: str | None = Field(
        default=None,
        min_length=1,
        max_length=4000,
        description="Optional structured prompt that bypasses automated generation.",
    )


class NegotiationGuidanceResponse(BaseModel):
    """AI-generated negotiation guidance returned to the caller."""

    model_config = ConfigDict(extra="forbid")

    summary: str = Field(..., min_length=1, description="High-level negotiation summary.")
    fallback_recommendation: str = Field(
        ...,
        min_length=1,
        description="Recommended fallback if the target position is unattainable.",
    )
    talking_points: list[str] = Field(
        default_factory=list,
        description="Bulletized arguments or talking points to use in the negotiation.",
    )
    risk_callouts: list[str] = Field(
        default_factory=list,
        description="Specific risks to monitor or escalate during negotiations.",
    )
    confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Confidence score for the generated guidance between 0 and 1.",
    )
    cached: bool = Field(
        default=False,
        description="Flag indicating whether the result was served from an in-memory cache.",
    )
    model: str = Field(..., min_length=1, description="Gemini model identifier used for generation.")
    latency_ms: int = Field(..., ge=0, description="End-to-end latency in milliseconds.")
    documentation_url: str | None = Field(
        default=None,
        description="Optional link to supporting documentation or playbook guidance.",
    )
    generated_prompt: str | None = Field(
        default=None,
        description="Final structured prompt that was supplied to the model.",
    )

    @field_validator("talking_points", "risk_callouts")
    @classmethod
    def _prune_empty_entries(cls, value: list[str]) -> list[str]:
        return [item.strip() for item in value if item and item.strip()]